### PR TITLE
boards: arm: Add support for nucleo_g071rb board

### PR DESCRIPTION
**Pull request by FRASTM**
This patch adds support for the STM32 nucleo_g071rb board from STMicroelectronics.

This patch requires  zephyrproject-rtos/hal_stm32#3

Signed-off-by: Francois Ramu <francois.ramu@st.com>

### CODE REVIEW

**Comment by FRASTM** [comment]
recheck

**Comment by FRASTM** [comment]
recheck

**Review by dbkinder — APPROVED** [review]
docs LGTM, thanks!

File: west.yml [hunk 1]
```diff
@@ -51,7 +51,7 @@ manifest:
       revision: b6c9262eed68000e69c92bef6e820cdbd5b0a32b
       path: modules/hal/st
     - name: hal_stm32
-      revision: b454ad819f4f10cb82ce5eb1d7f709a680bc61c5
+      revision: pull/3/head
```
Comment (erwango) [inline west.yml]: pull/13/head

**Review by erwango — CHANGES_REQUESTED** [review]

File: west.yml (hunk 1 above)
Comment (erwango) [inline west.yml]: This ommit will  have to be removed

File: dts/bindings/flash_controller/st,stm32g0-flash-controller.yaml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+---
```
Comment (erwango) [inline dts/bindings/flash_controller/st,stm32g0-flash-controller.yaml]: Remove this line

File: dts/bindings/flash_controller/st,stm32g0-flash-controller.yaml [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+    This binding gives a base representation of the STM32 G0 Flash Controller
+
+inherits:
+    !include flash-controller.yaml
+
+properties:
+    compatible:
+      constraint: "st,stm32g0-flash-controller"
+
+...
```
Comment (erwango) [inline dts/bindings/flash_controller/st,stm32g0-flash-controller.yaml]: Remove this line

File: soc/arm/st_stm32/stm32g0/Kconfig.soc [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+# Copyright (c) 2019 ST Microelectronics
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+choice
+prompt "STM32G0x MCU Selection"
+depends on SOC_SERIES_STM32G0X
+
+config SOC_STM32G071RB
```
Comment (erwango) [inline soc/arm/st_stm32/stm32g0/Kconfig.soc]: SOC_STM32G071XX

File: soc/arm/st_stm32/stm32g0/Kconfig.soc [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+	bool "STM32G071RB"
```
Comment (erwango) [inline soc/arm/st_stm32/stm32g0/Kconfig.soc]: STM32G071XX

File: soc/arm/st_stm32/stm32g0/Kconfig.defconfig.stm32g071rb [hunk 6]
```diff
@@ -0,0 +1,22 @@
+# Kconfig - ST Microelectronics STM32G071RB MCU
+#
+# Copyright (c) 2019 Philippe Retornaz <philippe@shapescale.com>
+# Copyright (c) 2019 ST Microelectronics
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+if SOC_STM32G071RB
```
Comment (erwango) [inline soc/arm/st_stm32/stm32g0/Kconfig.defconfig.stm32g071rb]: Use `SOC_STM32G071XX`
We don't need to introduce additional information (RB) here. This would prevent code factorization

File: soc/arm/st_stm32/stm32g0/Kconfig.defconfig.stm32g071rb [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+	default "stm32g071xx"
+
+config NUM_IRQS
+	int
+	default 32
+
+if GPIO_STM32
+
+endif # GPIO_STM32
+endif # SOC_STM32G071RB
```
Comment (erwango) [inline soc/arm/st_stm32/stm32g0/Kconfig.defconfig.stm32g071rb]: SOC_STM32G071XX

File: soc/arm/st_stm32/stm32g0/dts_fixup.h [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+ * Copyright (c) 2019 STMicroelectronics
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/* SoC level DTS fixup file */
+
+#define DT_NUM_IRQ_PRIO_BITS	        DT_ARM_V6M_NVIC_E000E100_ARM_NUM_IRQ_PRIORITY_BITS
+
+#define DT_PWM_STM32_3_DEV_NAME		    DT_ST_STM32_PWM_40000400_PWM_LABEL
```
Comment (erwango) [inline soc/arm/st_stm32/stm32g0/dts_fixup.h]: PWM defines should be introduced in PWM related commit

File: soc/arm/st_stm32/stm32g0/soc_registers.h [hunk 9]
```diff
@@ -0,0 +1,14 @@
+/*
```
Comment (erwango) [inline soc/arm/st_stm32/stm32g0/soc_registers.h]: Remove this file

File: soc/arm/st_stm32/stm32g0/flash_registers.h [hunk 10]
```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright (c) 2019 Philippe Retornaz <philippe@shapescale.com>
```
Comment (erwango) [inline soc/arm/st_stm32/stm32g0/flash_registers.h]: Add this file when adding Flash support

File: drivers/flash/CMakeLists.txt [hunk 11]
```diff
@@ -26,6 +26,7 @@ if(CONFIG_CLOCK_CONTROL_STM32_CUBE)
   zephyr_sources_ifdef(CONFIG_SOC_SERIES_STM32F7X flash_stm32f7x.c)
```
Comment (erwango) [inline drivers/flash/CMakeLists.txt]: Have you been able to test flash driver?
(using sample/drivers/flash_shell)

File: drivers/clock_control/clock_stm32g0.c [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+#endif /* CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL */
+
+/**
+ * @brief Activate default clocks
+ */
+void config_enable_default_clocks(void)
+{
+#if defined(CONFIG_EXTI_STM32) || defined(CONFIG_USB_DC_STM32)
+	/* Enable System Configuration Controller clock. */
+	LL_APB2_GRP1_EnableClock(LL_APB2_GRP1_PERIPH_SYSCFG);
```
Comment (erwango) [inline drivers/clock_control/clock_stm32g0.c]: Is that valid? Looking to previous file, it seemed APB2 was not supported

File: dts/arm/st/g0/stm32g0.dtsi [hunk 13]
```diff
@@ -46,9 +46,33 @@
...
 
+		rcc: rcc@40021000 {
+			compatible = "st,stm32-rcc";
+			clocks-controller;
+			#clock-cells = <2>;
+			reg = <0x40021000 0x400>;
+			label = "STM32_CLK_RCC";
+		};
+
+		timers3: timers@40000400 {
```
Comment (erwango) [inline dts/arm/st/g0/stm32g0.dtsi]: Keep this node for PWM commit

File: soc/arm/st_stm32/stm32g0/dts_fixup.h [hunk 14]
```diff
@@ -14,4 +14,7 @@
 #define DT_TIM_STM32_3_CLOCK_BITS		DT_ST_STM32_TIMERS_40000400_CLOCK_BITS
 #define DT_TIM_STM32_3_CLOCK_BUS		DT_ST_STM32_TIMERS_40000400_CLOCK_BUS
 
+#define DT_FLASH_DEV_BASE_ADDRESS		DT_ST_STM32G0_FLASH_CONTROLLER_40022000_BASE_ADDRESS
```
Comment (erwango) [inline soc/arm/st_stm32/stm32g0/dts_fixup.h]: Keep this one for flash commit

File: drivers/clock_control/clock_stm32_ll_common.c [hunk 15]
```diff
@@ -54,7 +54,7 @@ static void config_bus_clk_init(LL_UTILS_ClkInitTypeDef *clk_init)
 	clk_init->APB1CLKDivider = apb1_prescaler(
 					CONFIG_CLOCK_STM32_APB1_PRESCALER);
 
-#ifndef CONFIG_SOC_SERIES_STM32F0X
+#if !defined (CONFIG_SOC_SERIES_STM32F0X) && !defined (CONFIG_SOC_SERIES_STM32G0X)
```
Comment (erwango) [inline drivers/clock_control/clock_stm32_ll_common.c]: If APB2 not supported, you also need to update:
`config CLOCK_STM32_APB2_PRESCALER` in `drivers/clock_control/Kconfig.stm32`

File: drivers/pinmux/stm32/pinmux_stm32.c [hunk 16]
```diff
@@ -107,15 +107,20 @@ static int enable_port(u32_t port, struct device *clk)
 
 static int stm32_pin_configure(int pin, int func, int altf)
 {
+#ifdef CONFIG_GPIO
```
Comment (erwango) [inline drivers/pinmux/stm32/pinmux_stm32.c]: This change should be done and explained in a separate PR

File: dts/arm/st/g0/stm32g0.dtsi [hunk 17]
```diff
@@ -106,6 +106,24 @@
...
+
+		usart2: serial@40004400 {
+			compatible = "st,stm32-usart", "st,stm32-uart";
+			reg = <0x40004400 0x400>;
+			clocks = <&rcc STM32_CLOCK_BUS_APB1 0x00020000>;
+			interrupts = <28 0>;
+			status = "disabled";
+			label = "UART_2";
+		};
+
```
Comment (erwango) [inline dts/arm/st/g0/stm32g0.dtsi]: We miss the lpuart node

File: soc/arm/st_stm32/stm32g0/dts_fixup.h [hunk 18]
```diff
@@ -55,13 +55,30 @@
...
+
+#define DT_UART_STM32_USART_2_BASE_ADDRESS	DT_ST_STM32_USART_40004400_BASE_ADDRESS
+#define DT_UART_STM32_USART_2_BAUD_RATE	DT_ST_STM32_USART_40004400_CURRENT_SPEED
+#define DT_UART_STM32_USART_2_IRQ_PRI	DT_ST_STM32_USART_40004400_IRQ_0_PRIORITY
+#define DT_UART_STM32_USART_2_NAME		DT_ST_STM32_USART_40004400_LABEL
+#define DT_USART_2_IRQ				DT_ST_STM32_USART_40004400_IRQ_0
+#define DT_UART_STM32_USART_2_CLOCK_BITS	DT_ST_STM32_USART_40004400_CLOCK_BITS
+#define DT_UART_STM32_USART_2_CLOCK_BUS	DT_ST_STM32_USART_40004400_CLOCK_BUS
+#define DT_UART_STM32_USART_2_HW_FLOW_CONTROL	DT_ST_STM32_USART_40004400_HW_FLOW_CONTROL
+
```
Comment (erwango) [inline soc/arm/st_stm32/stm32g0/dts_fixup.h]: Miss the lpuart

File: soc/arm/st_stm32/stm32g0/soc.h [hunk 19]
```diff
@@ -44,6 +44,11 @@
 #include <stm32g0xx_ll_gpio.h>
 #endif
 
+#ifdef CONFIG_SERIAL_HAS_DRIVER
+#include <stm32g0xx_ll_usart.h>
+#include <stm32g0xx_ll_lpuart.h>
```
Comment (erwango) [inline soc/arm/st_stm32/stm32g0/soc.h]: Either remove this line, either add the LPUART node support

File: boards/arm/nucleo_g071rb/doc/index.rst [hunk 20]
```diff
@@ -0,0 +1,186 @@
...
+| UART      | on-chip    | serial port-polling;                |
+|           |            | serial port-interrupt               |
++-----------+------------+-------------------------------------+
+| PINMUX    | on-chip    | pinmux                              |
++-----------+------------+-------------------------------------+
+| GPIO      | on-chip    | gpio                                |
++-----------+------------+-------------------------------------+
+| CLOCK     | on-chip    | reset and clock control             |
++-----------+------------+-------------------------------------+
+| I2C       | on-chip    | i2c controller                      |
```
Comment (erwango) [inline boards/arm/nucleo_g071rb/doc/index.rst]: Not available yet

File: boards/arm/nucleo_g071rb/doc/index.rst [hunk 21]
```diff
@@ -0,0 +1,186 @@
...
++-----------+------------+-------------------------------------+
+| SPI       | on-chip    | spi controller                      |
```
Comment (erwango) [inline boards/arm/nucleo_g071rb/doc/index.rst]: Not available yet

File: boards/arm/nucleo_g071rb/support/openocd.cfg [hunk 22]
```diff
@@ -0,0 +1,23 @@
+# This is an ST NUCLEO-G071RB board with single STM32G071RB chip.
+# http://www.st.com/en/evaluation-tools/nucleo-g071rb.html
+source [find interface/stlink.cfg]
+
+transport select hla_swd
+
+set WORKAREASIZE 0x2000
+
+source [find target/stm32g0.cfg]
```
Comment (erwango) [inline boards/arm/nucleo_g071rb/support/openocd.cfg]: Is that working?

**Review by erwango — CHANGES_REQUESTED** [review]

**Comment by FRASTM** [comment]
recheck

File: drivers/clock_control/clock_stm32_ll_common.c (hunk 15 above)
Comment (erwango) [reply drivers/clock_control/clock_stm32_ll_common.c]: This comment still stands

File: drivers/clock_control/clock_stm32g0.c (hunk 12 above)
Comment (erwango) [reply drivers/clock_control/clock_stm32g0.c]: This comment still stands

File: drivers/pinmux/stm32/pinmux_stm32.c (hunk 16 above)
Comment (erwango) [reply drivers/pinmux/stm32/pinmux_stm32.c]: This comment still stands

File: soc/arm/st_stm32/stm32g0/dts_fixup.h (hunk 8 above)
Comment (erwango) [reply soc/arm/st_stm32/stm32g0/dts_fixup.h]: Not sure you you move these in this commit

File: boards/arm/nucleo_g071rb/CMakeLists.txt [hunk 23]
```diff
@@ -0,0 +1,5 @@
+# SPDX-License-Identifier: Apache-2.0
+
+zephyr_library()
```
Comment (erwango) [inline boards/arm/nucleo_g071rb/CMakeLists.txt]: Need to protect under if(CONFIG_PINMUX).
Cf other boards following recent rebase

File: boards/arm/nucleo_g071rb/support/openocd.cfg (hunk 22 above)
Comment (erwango) [reply boards/arm/nucleo_g071rb/support/openocd.cfg]: Question stands

**Review by erwango — CHANGES_REQUESTED** [review]

**Comment by erwango** [comment]
@FRASTM , can you take into account upcoming https://github.com/zephyrproject-rtos/zephyr/pull/17122?

File: boards/arm/nucleo_g071rb/CMakeLists.txt (hunk 23 above)
Comment (FRASTM) [reply boards/arm/nucleo_g071rb/CMakeLists.txt]: done

File: soc/arm/st_stm32/stm32g0/dts_fixup.h [hunk 24]
```diff
@@ -0,0 +1,15 @@
...
+ * Copyright (c) 2019 STMicroelectronics
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/* SoC level DTS fixup file */
+
+#define DT_NUM_IRQ_PRIO_BITS	        DT_ARM_V6M_NVIC_E000E100_ARM_NUM_IRQ_PRIORITY_BITS
+
+#define DT_TIM_STM32_3_BASE_ADDRESS		DT_ST_STM32_TIMERS_40000400_BASE_ADDRESS
```
Comment (erwango) [inline soc/arm/st_stm32/stm32g0/dts_fixup.h]: Please remove these

**Review by erwango — CHANGES_REQUESTED** [review]

File: boards/arm/nucleo_g071rb/support/openocd.cfg (hunk 22 above)
Comment (FRASTM) [reply boards/arm/nucleo_g071rb/support/openocd.cfg]: uart:~$ flash erase 0x1E000 0x1000
uart:~$ flash write 0x1E000 00 01 02 03 04 05 06 07
Reading back written bytes:
00 01 02 03 | 04 05 06 07
uart:~$ flash read 0x1E000 10
00 01 02 03 | 04 05 06 07
ff ff

File: drivers/flash/CMakeLists.txt (hunk 11 above)
Comment (FRASTM) [reply drivers/flash/CMakeLists.txt]: Testing samples/drivers/flash_shell  --> PASSED

uart:~$ flash erase 0x1E000 0x1000
uart:~$ flash write 0x1E000 00 01 02 03 04 05 06 07
Reading back written bytes:
00 01 02 03 | 04 05 06 07
uart:~$ flash read 0x1E000 10
00 01 02 03 | 04 05 06 07
ff ff

**Review by erwango — APPROVED** [review]

**Review by erwango — CHANGES_REQUESTED** [review]
LGTM, but please fix checkpatch first (dts_fixup.h warnings could be omitted)

**Comment by FRASTM** [comment]
> LGTM, but please fix checkpatch first (dts_fixup.h warnings could be omitted)

done 
checkpatch — Checks passed

**Comment by FRASTM** [comment]
> > LGTM, but please fix checkpatch first (dts_fixup.h warnings could be omitted)
> 
> done
> checkpatch — Checks passed

**Review by erwango — APPROVED** [review]

**Review by nashif — APPROVED** [review]

File: drivers/clock_control/clock_stm32g0.c [hunk 25]
```diff
@@ -0,0 +1,51 @@
+/*
```
Comment (ydamigos) [inline drivers/clock_control/clock_stm32g0.c]: Look very similar to `clock_stm32l4_wb.c`. Is it possible to use `clock_stm32l4_wb.c` instead of `clock_stm32g0.c`?

File: drivers/clock_control/Kconfig.stm32g0 [hunk 26]
```diff
@@ -0,0 +1,51 @@
+# Kconfig - STM32G0 PLL configuration options
```
Comment (ydamigos) [inline drivers/clock_control/Kconfig.stm32g0]: Look very similar to `Kconfig.stm32l4_wb`. Is it possible to use `Kconfig.stm32l4_wb` instead of `Kconfig.stm32g0`?

Comment (FRASTM) [reply drivers/clock_control/Kconfig.stm32g0]: yes, could be with a "Kconfig.stm32l4_wb_g0"

Comment (erwango) [reply drivers/clock_control/Kconfig.stm32g0]: For consistency, I'd prefer we keep aligned Kconfig.stm32xx and clock_xx.c files.
So either we keep them split, either we merge both.

File: drivers/clock_control/clock_stm32g0.c (hunk 25 above)
Comment (ydamigos) [reply drivers/clock_control/clock_stm32g0.c]: @FRASTM From your comment below, it is possible to combine Kconfig.stm32l4_wb and Kconfig.stm32g0 in a  Kconfig.stm32l4_wb_g0. Could we do the same for clock_stm32l4_wb.c and clock_stm32g0.c?

**Review by MaureenHelm — APPROVED** [review]
